### PR TITLE
General vaults review and fixes

### LINKED ISSUE #260: A single digit modification is not detected by isomorphic-git's `status` call
### Describe the bug

When making a change in a file using `await vault.commit(...)` and changing only a single number on that file will sometimes cause isomorphic-git to not recognise that the file has been modified and therefore if no other changes were made a commit will not be made even though a file was changed.

### To Reproduce

1. Add a file using `await vault.commit`
```
      const secretVer1 = {
        name: secretName,
        content: 'Secret-1-content-ver1',
      };
```
2. Modify the file using `await vault.commit`
```
      const secretVer2 = {
        name: secretName,
        content: 'Secret-1-content-ver2',
      };
```
3. Check the log using `await vault.log` to see if the most recent modification shows up.

NOTE: This won't happen ALL the time. It seems to be most common when running large groups of tests together?

### Expected behavior

We expect any file change, even a single digit change to be detected by isomorphic-git and therefore create a new commit when `vault.commit` is used.

### Additional context

- Most likely related to this [issue](https://github.com/isomorphic-git/isomorphic-git/issues/1245) on `isomorphic-git` 
- This issue has given a possible workaround by using `resetIndex` on each file in the vault
- Discussion about the problem is [here](https://gitlab.com/MatrixAI/Engineering/Polykey/js-polykey/-/merge_requests/205#note_711060987)
- This [issue](https://github.com/MatrixAI/js-encryptedfs/issues/55) created for EFS has the potential to resolve this issue

### Resolved

This was addressed by doing the following.

1. All contents of the vault are added using `git.add` before processing the `statusMatrix`. Making sure everything is staged beforehand essentially solves the issue. 
2. Simplified the logic for handling the `statusMatrix` to a much simpler `switch` statement.

### LINKED ISSUE #714: Inconsistent vault log when pulling secrets from a modified vault (received vault log is different) - should use fast forward merge
### Describe the bug

When I pull the changes from a modified vault on my local system, the log on both machines were different, while they should be identical.

```
cmcdragonkai ➜ matrix-framework-13-ryzen-7040  ➜ ~/Projects/Polykey-CLI
 $ ./dist/polykey.js vaults log myvault
commitId     6641db528e0e79412898425a3ea84e349aa5c82b
committer    vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp    Fri May 03 2024 19:53:49 GMT+1000 (Australian Eastern Standard Time)
message      "abc modified\n"
commitId     4e648a16325ebedd0847473ec04f02a0891b1be1
committer    vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp    Fri May 03 2024 19:45:41 GMT+1000 (Australian Eastern Standard Time)
message      "abc modified\n"
commitId     c987ff12719344b962e9586aa25b7a28bb328bc8
committer    vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp    Fri May 03 2024 19:45:13 GMT+1000 (Australian Eastern Standard Time)
message      "abc added\n"
commitId     88e97c3937ac2ee86595918d5ef136b9b92d0da0
committer    vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp    Fri May 03 2024 19:44:02 GMT+1000 (Australian Eastern Standard Time)
message      "Initial Commit\n"
```

```
aryanj@supercomputer ❯ ./pk vaults log myvault                                                       
commitId        aa08262fd12acc954561093cd1d4784cf772d588
committer       vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp       Fri May 03 2024 19:54:14 GMT+1000 (Australian Eastern Standard Time)
message         "Merge branch \'HEAD\' of http://\n"
commitId        6641db528e0e79412898425a3ea84e349aa5c82b
committer       vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp       Fri May 03 2024 19:53:49 GMT+1000 (Australian Eastern Standard Time)
message         "abc modified\n"
commitId        dd0d131b9997c74cabfe7202e68d3044fe59a5de
committer       vet1dhoduhkmn4749r8usiopivklr8i4qoh9kjnhtgeg5rie4nvh0
timestamp       Fri May 03 2024 19:48:42 GMT+1000 (Australian Eastern Standard Time)
message         "Initial Commit\n"
commitId        4e648a16325ebedd0847473ec04f02a0891b1be1
committer       vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp       Fri May 03 2024 19:45:41 GMT+1000 (Australian Eastern Standard Time)
message         "abc modified\n"
commitId        c987ff12719344b962e9586aa25b7a28bb328bc8
committer       vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp       Fri May 03 2024 19:45:13 GMT+1000 (Australian Eastern Standard Time)
message         "abc added\n"
commitId        88e97c3937ac2ee86595918d5ef136b9b92d0da0
committer       vcj3ufun2506jn44q7rr9bc1hv4c6eebicusfj1bl9sr27ookknlg
timestamp       Fri May 03 2024 19:44:02 GMT+1000 (Australian Eastern Standard Time)
message         "Initial Commit\n"
```

### To Reproduce

1. Add a secret to a vault
2. Share the vault
3. Modify the secret in the vault
4. Pull changes for the log
5. Observe different logs

### Expected behavior

The logs should be identical for any and all instances of the vault, irrelevant of the vault pulling new changes or not.

### Platform (please complete the following information)

* Device: Acer Spin 3 (SP314-54N)
* OS: Arch Linux (Kernel 6.8.5-arch1-1 64-bit)
* Version: 1.2.3-alpha.4

### Additional context

> image
>
> This is a bug with the vault log - pull system. Needs to be fast forward merge.
> \_Originally posted by @CMCDragonkai in [https://github.com/MatrixAI/Polykey/issues/709#issuecomment-2095014382\_](https://github.com/MatrixAI/Polykey/issues/709#issuecomment-2095014382\_)

### Resolved

This was adressed by doing the following.

1. When calling `git.clone` and `git.pull` I made sure to target the `canonicalBranch` ref name. This really was all that was needed to make sure that the history was correct.
2. I also enabled the `fastForward` and `fastForwardOnly` flags for `git.pull` to ensure only linear history was pulled from the remote.

### LINKED ISSUE #718: Minor review and refactor of vaults domain for 'polykey'
### Specification

As part of refactoring the `git` domain in #298 we need to do a small review of the vaults domain for any fixes and improvements. This is a general refactor and review. We're looking for any places that can be tidied up if needed. Some tests need to be re-enabled as well.

### Additional context

* Related: #298
* Related #709
* Related #717
* Related #714
* Related #523

### Tasks

1. General review of the vaults domain.
2. Have a closer look at the vaults committing secrets and git garbage collection code.
3. Determine and apply any optimisations that can be made.
4. Review vaults tests and clean them up if needed.
5. Re-enable any tests that have been disabled.
6. Possibly split up some test files so the vaults domain runs quicker in CI. vault ops is currently very slow.

### PR DESCRIPTION
**Pull request by tegefaulkes**
### Description

This PR handles a general review of the vaults domain. General optimisations if they can be applied. Minor fixes for vault related issues and some touch ups to the vaults tests.

NOTE: this branches off from the git refactoring PR and is targeting that for now. But once the git refactoring is merged then this needs to rebase and target staging.

### Issues Fixed

* Ref ENG-316
* Fixes #718 
* Fixes #260 
* Fixes #714 

### Tasks

- [x] 1. Review vaults domain, clean up and refactor anything as needed
- [x] 2. Address #260 
- [x] 3. Address #714 
- [x] 4. Have a closer look at the vaults committing secrets and git garbage collection code.
- [x] 5. Determine and apply any optimisations that can be made.
- [x] 6. Review vaults tests and clean them up if needed.
- [x] 7. Re-enable any vaults tests that have been disabled.
- [x] 8. Possibly split up some test files so the vaults domain runs quicker in CI. vault ops is currently very slow.

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
What's the order of merging in relation to #709?

**Comment by tegefaulkes** [comment]
#709 merges first and then this one.

**Comment by tegefaulkes** [comment]
Switched base over staging now that git refactoring has been merged.

**Comment by tegefaulkes** [comment]
In the `pullVault` function in `VaultInternal`. There is an `let error` variable that is never set or modified but it affects some logic. I need to dig into this since it can't be intended as it is.

**Comment by tegefaulkes** [comment]
One of the vaults test for pulling was failing randomly. With some testing I found that it's random network traffic with the mainnet.

image


I'll need to make sure any tests using a node connection don't actually try to join the mainnet when testing. In this case we have a `NodeConnectionManager` for testing but give it an empty manifest. On the same note, it's very odd that the test is failing for failing to handle incoming RPC requests. This shouldn't crash us so maybe I need to look deeper into that.

So yeah, there are two problems here. 
1. I have a minimal `NodeConnectionManager` needed to start forward connections but it's receiving reverse connections from the mainnet. Likely from trying to join it itself. We shouldn't be starting or receiving external connections unless the test explicitly wants to do this. 
2. There is an `RPC` error for a missing handler propagating outwards causing the test to fail. By design this should only bubble through the stream back through the `RPC`. This isn't a critical error that should ever leave the `RPC`. Something is wrong here.

**Comment by tegefaulkes** [comment]
I may need to add #461 to this to address the intermittent failure mentioned above. Though at the same time, I feel like this PR is getting too large in scope? I'll think about it.

**Comment by tegefaulkes** [comment]
> One of the vaults test for pulling was failing randomly. With some testing I found that it's random network traffic with the mainnet.
> 
> image
> 
> I'll need to make sure any tests using a node connection don't actually try to join the mainnet when testing. In this case we have a `NodeConnectionManager` for testing but give it an empty manifest. On the same note, it's very odd that the test is failing for failing to handle incoming RPC requests. This shouldn't crash us so maybe I need to look deeper into that.
> 
> So yeah, there are two problems here.
> 
> 1. I have a minimal `NodeConnectionManager` needed to start forward connections but it's receiving reverse connections from the mainnet. Likely from trying to join it itself. We shouldn't be starting or receiving external connections unless the test explicitly wants to do this.
> 2. There is an `RPC` error for a missing handler propagating outwards causing the test to fail. By design this should only bubble through the stream back through the `RPC`. This isn't a critical error that should ever leave the `RPC`. Something is wrong here.

I made a new issue #722 And it will be addressed later, outside this PR.

**Comment by tegefaulkes** [comment]
`vaultOps.addSecretDirectory` is a weird case. All other vaultOps operate purely within the EFS with an secret contents passed directly into it.  but `addSecretDirectory` seems to access the `fs` directly reading from it to copy a directory into a vault. 

I don't like this. It makes it the only secret command that functions against the node's filesystem and not the `Polykey-CLI`'s file system. We need to review this and work out what we want to do with it.

**Comment by tegefaulkes** [comment]
So `addSecretDirectory` needs to be refactored in how it works. currently it accesses the fs to read the directory when copying it to the vault. However, as mentioned in all other vaultOps that write data, the data is passed into the function. From the `CLI` perspective, the files are sent over the RPC when adding to the vault.

The problem here is `addSecretDirectory` is the odd duck out. The data isn't sent over the RPC, it's read directly from the `fs`. So we need to refactor it. There are two options

1. We refactor it so the directory and all it's contents are sent over the RPC to the `addSecretDirectory` command. A bit of a hassle but do-able. We'd have to encode directory and file contents in a binary stream. I'd rather handle this in a separate PR.
2. We remove the `addSecretDirectory` and it's `CLI` command. Seems like a QOL function.

**Comment by tegefaulkes** [comment]
I'm working on #714 right now. It's a pretty simple fix. However I found that we don't test the case where you pull a vault that deviated from the original history.

I modified a test to check this and `isomorphic-git` doesn't actually support handling that.

```
Merges with conflicts are not supported yet.
ErrorVaultsMergeConflict: Merges with conflicts are not supported yet.
    at constructor_.pullVault (/home/brian/workspace/Polykey/src/vaults/VaultInternal.ts:651:15)
    at /home/brian/workspace/Polykey/src/vaults/VaultManager.ts:789:9
    at withF (/home/brian/workspace/Polykey/node_modules/@matrixai/resources/src/utils.ts:24:12)
    at constructor_.pullVault (/home/brian/workspace/Polykey/src/vaults/VaultManager.ts:784:5)
    at withF (/home/brian/workspace/Polykey/node_modules/@matrixai/resources/src/utils.ts:24:12)
    at Object.<anonymous> (/home/brian/workspace/Polykey/tests/vaults/VaultManager.test.ts:1138:9)
```

So I'm not actually sure it's possible to handle this case. We'd have to clone the vault from scratch to handle it really. Do we just throw an error urging the user to clone the vault fresh due to the deviating history?

**Comment by CMCDragonkai** [comment]
When you a pull a vault into a deviated history. This is what supposed to happen:

1. You are not allowed to do it.

At a higher level, the user upon editing a vault that is "pulled", should be doing a COW operation, the vault must be duplicated and separated from the old vault. The old vault maintains its ability to be pulled.

This means you cannot have a deviated history at all. It's not allowed by the constraints of the system.

**Comment by tegefaulkes** [comment]
I'm aware that we restrict editing of a vault that has been cloned. But the remote vault we clone from can do a branching commit and completely change it's history. Do we apply the same restraint in this case? Where we disallow pulling a vault who's history no longer matches our cloned vault's history?

Different sides of the same coin really. Just want to be clear since the example you gave was the opposite side to the example I described.

**Comment by CMCDragonkai** [comment]
> I'm aware that we restrict editing of a vault that has been cloned. But the remote vault we clone from can do a branching commit and completely change it's history. Do we apply the same restraint in this case? Where we disallow pulling a vault who's history no longer matches our cloned vault's history?
> 
> Different sides of the same coin really. Just want to be clear since the example you gave was the opposite side to the example I described.

It must be disallowed on both sides.

Linear history must be maintained. If the other side corrupted the vault, you just decline the ability to pull. You indicate why.

**Comment by tegefaulkes** [comment]
I still need to do one last once-over, but it's ready for review. Doesn't need a deep review. Just so you are aware of changes and the more eyes on it the better. @CMCDragonkai @amydevs 

I'll run through the checklist after review. No point doing most of the checklist if there are changes that need to be made still.

File: src/vaults/VaultInternal.ts [hunk 1]
```diff
@@ -953,12 +953,12 @@ class VaultInternal {
...
+          ["c.txt", 0, 2, 3], // added, staged, with unStaged changes
           ["d.txt", 1, 1, 1], // unmodified
-          ["e.txt", 1, 2, 1], // modified, unstaged
+          ["e.txt", 1, 2, 1], // modified, unStaged
           ["f.txt", 1, 2, 2], // modified, staged
-          ["g.txt", 1, 2, 3], // modified, staged, with unstaged changes
-          ["h.txt", 1, 0, 1], // deleted, unstaged
+          ["g.txt", 1, 2, 3], // modified, staged, with unStaged changes
+          ["h.txt", 1, 0, 1], // deleted, unStaged
           ["i.txt", 1, 0, 0], // deleted, staged
```
Comment (CMCDragonkai) [inline src/vaults/VaultInternal.ts]: You might want to align your case sensitivity here.

Also should `Type StatusRow` be an actual type somewhere in `types.ts`?

File: src/vaults/VaultManager.ts [hunk 2]
```diff
@@ -806,7 +806,8 @@ class VaultManager {
     tran?: DBTransaction,
   ): AsyncGenerator<Buffer, void, void> {
     if (tran == null) {
-      const handleInfoRequest = (tran) => this.handleInfoRequest(vaultId, tran);
+      const handleInfoRequest = (tran: DBTransaction) =>
+        this.handleInfoRequest(vaultId, tran);
       return yield* this.db.withTransactionG(async function* (tran) {
         return yield* handleInfoRequest(tran);
       });
```
Comment (CMCDragonkai) [inline src/vaults/VaultManager.ts:813]: This seems rather pointless, it's used once. Maybe just inline it?

Comment (CMCDragonkai) [reply src/vaults/VaultManager.ts:813]: Also later it says `const efs = this.efs;` also seems pointless, make the code simpler... no need for internal assignment.

File: src/vaults/utils.ts [hunk 3]
```diff
@@ -55,8 +55,8 @@ async function* readDirRecursively(
   fs,
   dir = '.',
 ): AsyncGenerator<string, void, void> {
-  const dirents = await fs.promises.readdir(dir);
-  for (const dirent of dirents) {
+  const dirEntities = await fs.promises.readdir(dir);
+  for (const dirent of dirEntities) {
```
Comment (CMCDragonkai) [inline src/vaults/utils.ts]: I believe this is `dirEntries` and `dirEntry`.

File: tests/vaults/VaultOps.test.ts [hunk 4]
```diff
@@ -476,7 +476,7 @@ describe('VaultOps', () => {
       expect(secretList).toHaveLength(0);
     });
   });
-  // Not sure if hidden file names are a special case but I'm keeping the tests just in case
+  // Not sure if hidden file names are a special case, but I'm keeping the tests just in case
   test('adding hidden files and directories', async () => {
```
Comment (CMCDragonkai) [inline tests/vaults/VaultOps.test.ts]: Let's actually test this, and remove the comment.

**Comment by CMCDragonkai** [comment]
How was #260 and #714 fixed? Can you describe the solution in the spec?

**Comment by CMCDragonkai** [comment]
> > I'm aware that we restrict editing of a vault that has been cloned. But the remote vault we clone from can do a branching commit and completely change it's history. Do we apply the same restraint in this case? Where we disallow pulling a vault who's history no longer matches our cloned vault's history?
> > Different sides of the same coin really. Just want to be clear since the example you gave was the opposite side to the example I described.
> 
> It must be disallowed on both sides.
> 
> Linear history must be maintained. If the other side corrupted the vault, you just decline the ability to pull. You indicate why.

And do you have a corresponding exception for these cases now?

File: src/vaults/VaultManager.ts (hunk 2 above)
Comment (tegefaulkes) [reply src/vaults/VaultManager.ts:813]: Generic generators don't handle the `this` context very well compared to a normal arrow function. Anything that makes use of `this` in a generator needs to be put into a variable or an arrow function outside the generator for it to work properly. We do it in a bunch of places.

File: src/vaults/VaultInternal.ts (hunk 1 above)
Comment (tegefaulkes) [reply src/vaults/VaultInternal.ts]: The type is defined by `isomorphic-git` and we don't use it in any form besides consuming it in this one part of the code. It is only named in this comment block to make it clear what we're getting and how to interpret it.

File: src/vaults/utils.ts (hunk 3 above)
Comment (tegefaulkes) [reply src/vaults/utils.ts]: Fixed.

File: tests/vaults/VaultOps.test.ts (hunk 4 above)
Comment (tegefaulkes) [reply tests/vaults/VaultOps.test.ts]: Comment has been removed.

**Comment by tegefaulkes** [comment]
> > > I'm aware that we restrict editing of a vault that has been cloned. But the remote vault we clone from can do a branching commit and completely change it's history. Do we apply the same restraint in this case? Where we disallow pulling a vault who's history no longer matches our cloned vault's history?
> > > Different sides of the same coin really. Just want to be clear since the example you gave was the opposite side to the example I described.
> > 
> > 
> > It must be disallowed on both sides.
> > Linear history must be maintained. If the other side corrupted the vault, you just decline the ability to pull. You indicate why.
> 
> And do you have a corresponding exception for these cases now?

Yes, there was an existing error for this called `ErrorVaultsMergeConflict`

File: src/vaults/VaultManager.ts (hunk 2 above)
Comment (CMCDragonkai) [reply src/vaults/VaultManager.ts:813]: Don't do the binding pattern instead?

File: src/vaults/VaultInternal.ts (hunk 1 above)
Comment (CMCDragonkai) [reply src/vaults/VaultInternal.ts]: No I mean `unStaged` vs `unstaged`.

**Comment by CMCDragonkai** [comment]
Does `ErrorVaultsMergeConflict` work both ways? It generically means both conflict during a pull and any attempt to conflict it locally? Or are you just ensuring that one cannot corrupt one's own vault automatically?

Are you preventing the node from mutating its own (pulled/cloned) vault, and breaking history for a vault that has been pulled or cloned...? Or maybe the error only shows up on the pulling/cloning side?

**Comment by CMCDragonkai** [comment]
You should update the spec of this PR describing any protocol changes - and how was the solution implemented in a human readable manner. Including in reference to the above comments.

File: src/vaults/VaultManager.ts (hunk 2 above)
Comment (tegefaulkes) [reply src/vaults/VaultManager.ts:813]: As a personal rule I'll only use `.bind` if I absolutely have to. In this case I think it's cleaner to still treat it as a closure but to assign what we need to a variable just before it. it *looks* unnecessary, but its still very clear what's happening.

**Comment by tegefaulkes** [comment]
> Does `ErrorVaultsMergeConflict` work both ways? It generically means both conflict during a pull and any attempt to conflict it locally? Or are you just ensuring that one cannot corrupt one's own vault automatically?
> 
> Are you preventing the node from mutating its own (pulled/cloned) vault, and breaking history for a vault that has been pulled or cloned...? Or maybe the error only shows up on the pulling/cloning side?

Yes, weather the local history diverged or the remove diverged, it looks like a disconnect in the history that requires a merge. So the same error in both cases.

We can't modify local history anyway since any attempt to do so will result in a `ErrorVaultRemoteDefined: Vault is a clone of a remote vault and can not be mutated` error. So cloned vaults are immutable.

File: src/vaults/VaultManager.ts (hunk 2 above)
Comment (CMCDragonkai) [reply src/vaults/VaultManager.ts:813]: I think `const efs = this.efs;` is however useless though right?

**Comment by tegefaulkes** [comment]
Ready to merge. I'll write up the changes in a moment.

File: src/git/http.ts [hunk 5]
```diff
@@ -398,12 +399,18 @@ async function* generatePackData({
...
+      dir,
+      gitdir: gitDir,
+      oids: objectIds,
+    });
+  } catch (e) {
+    // Return without sending any data
+    return;
+  }
   if (packFile.packfile == null) utils.never('failed to create packFile data');
   let packFileBuffer = Buffer.from(packFile.packfile.buffer);
```
Comment (CMCDragonkai) [inline src/git/http.ts]: The `never` should be reserved for programming errors. If it can happen, it should be an exception.

Comment (tegefaulkes) [reply src/git/http.ts]: I mis-remembered, the `never` here is just to enforce the fact that `packfile` is defined. By our usage it should always be defined. So the `never` is proper here, just the message is off point.

```ts
    /**
     * - The packfile contents. Not present if `write` parameter was true, in which case the packfile was written straight to disk.
     */
    packfile?: Uint8Array;
};
```

**Comment by CMCDragonkai** [comment]
Ok message should be updated.

20 May 2024 14:36:00 Brian Botha ***@***.***>:

File: src/git/http.ts (hunk 5 above)
Comment (tegefaulkes) [reply src/git/http.ts]: I made some comments in the code to clarify things, Also removed the never for a `return` to align with the above catch block.

**Pull request merged**